<a href="https://colab.research.google.com/github/111Vidya/Python_fundamentals/blob/main/Mar_22_Naive_Bayes_Spam_classifier.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Thu Nov  7 18:11:36 2019

@author: srinivasgurrala
"""
!pip install joblib
import pandas as pd
import numpy as np
from sklearn.feature_extraction.text import CountVectorizer #convert text to numbers
from sklearn.model_selection import train_test_split # split data to train and test
from sklearn.naive_bayes import MultinomialNB # Naive Bayes library
from sklearn.metrics import classification_report

df = pd.read_csv('spam.csv', encoding="latin-1")
names=['class','message'] # Data already has col names .this fn can be ignored
df.columns=names
df['label'] = df['class'].map({'ham': 0, 'spam': 1})
X = df['message'] # Input
y = df['label'] # output
cv = CountVectorizer() # Intialize countvector
X = cv.fit_transform(X) # Fit and transform the Data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=42)
#Naive Bayes Classifier
clf = MultinomialNB()
clf.fit(X_train,y_train)
clf.score(X_test,y_test)
y_pred = clf.predict(X_test) # Predict
print(classification_report(y_test, y_pred))

import joblib # Ignore for now below fn
joblib.dump(clf, 'NB_spam_model.pkl')

NB_spam_model = open('NB_spam_model.pkl','rb')
clf = joblib.load(NB_spam_model)


              precision    recall  f1-score   support

           0       0.99      0.99      0.99      1587
           1       0.93      0.92      0.92       252

    accuracy                           0.98      1839
   macro avg       0.96      0.95      0.96      1839
weighted avg       0.98      0.98      0.98      1839



99% accuracy to identify ham(0)
92% accuracy to identify Spam(1)
Over all avg accuracy 98%